misc,ext,tests: Automatically split CI TestLib tests across GitHub Action jobs

BobbyRBruce opened the pull request:
This PR utilizes GitHub Action's matrix's to automatically distribute the CI testlib gem5 build and test jobs across available GitHub Action Runners.

The CI tests (the `quick` testlib tests, i.e. those run with `./main.py run`) are distributed across the runners on a per directory basis --- all directories under "tests/gem5" are run as their own jobs. 

The necessary gem5 builds for each workflow are now automatically inferred via the introduction of `./main.py list`'s  `--build-targets` flag which returns the gem5 build target for a given test or collection of tests. E.g., `./main.py list --build-targets` will return the build targets for all the `quick` testlib tests and `./main.py list --build-target --uid=<id>` will return the build targets the test suite `<id>` requires.

Moving from monolithic jobs to fine-grained ones will make the locaiton of test failures more obvious. Each job has it's own artifact containing "test/testing-results" for the tests run in that job. In addition, maintenance of these files should become less burdensome due to less hardcoding.

BobbyRBruce committed:
misc,tests: Split testlib CI Tests to one dir-per-job

This splits the CI Tests to one job per sub-directory in "tests/gem5"
via a matrix.

Advantages:
* We can utilize more runners to run the quick tests. This should mean
  tests run quicker.
* This approach does not require editing of the workflow as more tests
  are added or taken away.
* There is now an output artifact for each directory in "tests/gem5"
  instead of one for the entriety of every quick test in "tests".

In addition:
* The artifact retention for the test outputs has been increased to 30 days.
* The output test artifacts have been renamed to be more descriptive of
  the job, run, attempt, directory run, and the status.
* The 'tar' step has been removed. GitHub's 'action/artifact' can handle
  directories.

Change-Id: I5b3132b424e3769d81d9cd75db2a8c59dbe4a7e5

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -71,141 +62,85 @@ jobs:
+          # Therefore, we need to make the gem5 binaries executable.
+        - name: Chmod gem5.{opt,debug,fast} to be executable
+          run:
+            find . -name "gem5.opt" | xargs chmod u+x
```
nit, personally I'd use the following, but `xargs` works, so no need to change.

```suggestion
            find . -name "gem5.opt" -exec chmod u+x {} \;
```

BobbyRBruce replied:
Should be in the latest push.

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -71,141 +62,85 @@ jobs:
+           suite_ids=$(./main.py list --length=long | grep "^SuiteUID:")
+           {
+               echo 'SUITE_IDS<<EOF'
+               ./main.py list --length=long | grep "^SuiteUID:"
```
Would it be possible to specify the binary/variant that is needed in the runner step if we extended `main.py` to be able to output the right information? 

More generally, we might as well modify this script, or make a new one, that gives us the information we need here instead of using grep.

BobbyRBruce replied:
My latest push includes this. It's not a perfect solution but I think it works.

powerjg reviewed (commented):
This is *really cool*. Thanks for putting this together! One small nit and a suggestion below.

BobbyRBruce committed:
ext,tests: Fix `--figures` flag when using `./main.py list`

Now the "tests/main.py" script will accept the `--fixtures` flag when
using the `list` command. This will only list the fixtures needed.

To have this implemented `__str__` for the `Fixture` class has been
implemented.

Change-Id: I4bba26e923c8b0001163726637f2e48c801e92b1

BobbyRBruce committed:
ext,tests: Allow passing of `--uid` to `./main.py list`

This is useful for listing the fixtures of a Suite.

Change-Id: Id2f1294cc7dea03a6b26e8abc5083886fe0299d9

BobbyRBruce committed:
tests: Replace print with testlib.log for PARSEC warn

Using just a print was causing this warning to print even with the `-q`
flag was passed. The `-q` flag sets the output to machine readable,
which the warning statement is not.

Change-Id: I139e2565dbc53aaee9027c0e003d34ba800a7ef4

BobbyRBruce committed:
ext,tests: Add `--build-targets` option to `./main.py list`

This allows for build target information (i.e., the gem5 binary to be
built for the tests) to be returned.

Change-Id: I6638b54cbb1822555f58e74938d36043c11108ba

powerjg commented on .github/workflows/ci-tests.yaml:
```diff
@@ -80,39 +59,104 @@ jobs:
+      - uses: actions/checkout@v3
+      - name: Build gem5
+        run: scons ${{ matrix.build-target }} -j $(nproc)
+      - run: echo "temp" > anchor.txt
```
What is this file used for?

powerjg commented on .github/workflows/ci-tests.yaml:
```diff
@@ -80,39 +59,104 @@ jobs:
-      - run: chmod u+x build/ALL/gem5.opt
-      - name: The TestLib CI Tests
+          name: ci-tests-${{ github.run_number }}-attempt-${{ github.run_attempt }}-testlib-quick-all-gem5-builds
+      - run: ls -l
+      - run: find . -name "gem5.*"
```
For debugging?

BobbyRBruce replied:
yes. Not ready for review just yet

powerjg commented on .github/workflows/ci-tests.yaml:
```diff
@@ -80,39 +59,104 @@ jobs:
+          find . -name "gem5.fast" -exec chmod u+x {} \;
+
+        # Run the testlib quick tests in the given directory.
+      - name: Run "tests/${{ matrix.test-dir }}"" TestLib quick tests
```
Is there an extra `"` here?

BobbyRBruce committed:
tests: Remove ":" from testing results output dir name

Colons in path names is not advisable.

Change-Id: I7748a36cabafde69759f7a9892f7b8910470b85e

powerjg commented:
BTW, this is really cool, still! Is there anything you need from me to make progress?

BobbyRBruce committed:
python: Recursively create checkpoint dir

While there was code present in "serialize.cc" to create the checkpoint
directory, it did not do recursively. This patch ensures all the
directories are created in a path to the checkpoint directory.

Change-Id: Ibcf7f800358fd89946f550b8cfb0cef8b51fceac

BobbyRBruce commented:
@powerjg and @mkjost0 : this is **finally** ready for review... far too much trial and error.

You can see the latest test passing as evidence of it running well: https://github.com/gem5/gem5/actions/runs/6178661302. You can download the artifacts for any directory under "tests/gem5".

At a quick glance it seems like this is roughly 40 minutes faster. the asmtests are the weakest point, they take an hour. If we really wanted to do better we could split these up but it's good for now.

powerjg commented on tests/gem5/fixture.py:
```diff
@@ -74,7 +76,7 @@ def post_test_procedure(self, testitem):
         testUID = testitem.metadata.name
         testing_result_folder = os.path.join(
-            config.result_path, "SuiteUID:" + suiteUID, "TestUID:" + testUID
+            config.result_path, "SuiteUID-" + suiteUID, "TestUID-" + testUID
```
This might break how people run tests in batches. We should make sure that @giactra and @abmerop are aware and don't have any big issues.

BobbyRBruce replied:
Just to clarify, this is only changing the name of the directories in "tests/testing-results". For some reason when I created this code I used a colon. This isn't allowed in file/directory names in some OS's (MS Windows at least), and needs updated here as the `action/artifact-upload` and `action/artifact-download` actions don't allow them in file/directory names.

What has not changed is the format of the SuiteIDs (i.e., that you pass to `--uid` when running a specific job). This is just a file name fix.

powerjg reviewed (commented):
lgtm. I think this is a *massive* improvement. Thanks, Bobby!
